fabric-transfer-api-v1/src/main/java/net/fabricmc/fabric/api/transfer/v1/fluid/FluidVariantProperties.java

```diff
@@ -0,0 +1,167 @@
+
+package net.fabricmc.fabric.api.transfer.v1.fluid;
+
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Nullable;
+
+import net.minecraft.fluid.Fluid;
+import net.minecraft.fluid.Fluids;
+import net.minecraft.sound.SoundEvent;
+import net.minecraft.sound.SoundEvents;
+import net.minecraft.text.Text;
+import net.minecraft.world.World;
+
+import net.fabricmc.fabric.api.lookup.v1.custom.ApiProviderMap;
+
+/**
+ * Common fluid variant properties, accessible both client-side and server-side.
+ *
+ * <p><b>Experimental feature</b>, we reserve the right to remove or change it without further notice.
+ * The transfer API is a complex addition, and we want to be able to correct possible design mistakes.
+ */
+@ApiStatus.Experimental
+public class FluidVariantProperties {
+	private static final ApiProviderMap<Fluid, FluidVariantPropertyHandler> HANDLERS = ApiProviderMap.create();
+	private static final FluidVariantPropertyHandler DEFAULT_HANDLER = new FluidVariantPropertyHandler() { };
+
+	/**
+	 * Register a property handler for the passed fluid.
+	 */
+	public static void register(Fluid fluid, FluidVariantPropertyHandler handler) {
+		if (HANDLERS.putIfAbsent(fluid, handler) != null) {
+			throw new IllegalArgumentException("Duplicate handler registration for fluid " + fluid);
+		}
+	}
+
+	/**
+	 * Return the property handler for the passed fluid, if available, and {@code null} otherwise.
+	 */
+	@Nullable
+	public static FluidVariantPropertyHandler getHandler(Fluid fluid) {
+		return HANDLERS.get(fluid);
+	}
+
+	/**
+	 * Return the property handler for the passed fluid, if available, or the default instance otherwise.
+	 */
+	public static FluidVariantPropertyHandler getHandlerOrDefault(Fluid fluid) {
+		FluidVariantPropertyHandler handler = HANDLERS.get(fluid);
+		return handler == null ? DEFAULT_HANDLER : handler;
+	}
+
+	/**
+	 * Return the name that should be used for the passed fluid variant.
+	 */
+	public static Text getName(FluidVariant variant) {
+		return getHandlerOrDefault(variant.getFluid()).getName(variant);
+	}
+
+	/**
+	 * Return the sound corresponding to a container of this fluid variant being filled.
+	 */
+	public static SoundEvent getFillSound(FluidVariant variant) {
+		return getHandlerOrDefault(variant.getFluid()).getFillSound(variant);
+	}
+
+	/**
+	 * Return the sound corresponding to a container of this fluid variant being emptied.
+	 */
+	public static SoundEvent getEmptySound(FluidVariant variant) {
+		return getHandlerOrDefault(variant.getFluid()).getEmptySound(variant);
+	}
+
+	/**
+	 * Return an integer in [0, 15]: the light level emitted by this fluid variant, or 0 if it doesn't naturally emit light.
+	 */
+	public static int getLuminance(FluidVariant variant) {
+		int luminance = getHandlerOrDefault(variant.getFluid()).getLuminance(variant);
+
+		if (luminance < 0 || luminance > 15) {
+			throw new RuntimeException("Invalid luminance %d for fluid variant %s".formatted(luminance, variant));
```

Comment:
We should probably throw `IllegalStateException` and mention that the handler is broken as well.

Reply:
I don't think ISE works: `Signals that a method has been invoked at an illegal or inappropriate time.`

I now mention that the handler is broken.